### ISSUE TITLE
`Leaves` optic

### ISSUE THREAD

**Issue by cscherrer** [JuliaObjects/Accessors.jl#14]
@jw3126 I think this is working well now. What do you think of moving `Leaves` into Accessors? It's very fast, and seems it could be useful for others.

Most of the code is [here](https://github.com/cscherrer/NestedTuples.jl/blob/master/src/leaves.jl). Biggest concern I could see is that it depends on GeneralizedGenerated. There may be a way to drop that dependency, but it would take some fiddling.

It's no trouble to me either way, and I won't be offended if you don't think it's a good fit. Just seems more natural to me to have it in Accessors.

**Comment by jw3126** [comment]
Nice! Accessing the leaves of a tree is certainly an important task. As I understand, this assumes that the tree structure is encoded using `Tuples` and `NamedTuples`. There are many other possible choices, however. What if I have a tree like this:
```julia
["some", ["tree", [["encoded", "using"], "vectors"]
```
Is there a way to let the user supply the tree structure?

**Comment by cscherrer** [comment]
Hmm, good question. It's not clear to me in this case even what we'd want the result to look like. Julia infers the type as `Array{Any, 1}`, so that doesn't help.

OTOH, if we instead had some nested structs, that should be manageable. StructArrays has some nice tricks for recasting a struct as a named tuple, and after that it reduces to the previous case.

My interest here was to just work with the part that can be manipulated very quickly. So one approach could be to use this as an "outer optic" and pass to something else for each resolved component.

Then I guess part of the question is whether the current results should be considered "leaves". Maybe `TupleLeaves` or something is a better name?

**Comment by jw3126** [comment]
> Hmm, good question. It's not clear to me in this case even what we'd want the result to look like. Julia infers the type as `Array{Any, 1}`, so that doesn't help.

I would expect bad performance in this case, but the semantics would be:
```julia
tree = ["some", ["tree", [["encoded", "using"], "vectors"]
modify(uppercase, tree, Leaves(...)) == ["SOME", ["TREE", [["ENCODED", "USING"], "VECTORS"]
```
My point was that there are many ways trees are encoded and instead of only supporting one somewhat arbitrary way (descent into `Tuples` and `NamedTuples`) I would prefer if Accessors.jl could be helpful for all these cases.

> My interest here was to just work with the part that can be manipulated very quickly. So one approach could be to use this as an "outer optic" and pass to something else for each resolved component.

I think [Recursive](https://github.com/JuliaObjects/Accessors.jl/blob/4e839837c29337650e0c8bc8c9e92a54445289c8/src/optics.jl#L276-L295) goes in this direction. Let's hope inference can handle it :smile:

**Comment by jw3126** [comment]
```julia
using Accessors
x = (a = [1, 2], b = (c = ([3, 4], [5, 6]), d = [7, 8]));
optic = Recursive(x -> x isa Union{Tuple, NamedTuple}, Elements())
@test modify(sum, x, optic, ) === (a = 3, b = (c = (7, 11), d = 15))
```
But sadly
```julia
@code_warntype modify(sum, x, optic, )
@code_warntype modify(sum, x, optic, )

Variables
  #self#::Core.Compiler.Const(Accessors.modify, false)
  f::Core.Compiler.Const(sum, false)
  obj::NamedTuple{(:a, :b),Tuple{Array{Int64,1},NamedTuple{(:c, :d),Tuple{Tuple{Array{Int64,1},Array{Int64,1}},Array{Int64,1}}}}}
  optic::Core.Compiler.Const(Recursive{var"#3#4",Elements}(var"#3#4"(), Elements()), false)

Body::NamedTuple{(:a, :b),_A} where _A<:Tuple
1 ─      nothing
│   %2 = Accessors.OpticStyle(optic)::Core.Compiler.Const(Accessors.ModifyBased(), false)
│   %3 = Accessors._modify(f, obj, optic, %2)::NamedTuple{(:a, :b),_A} where _A<:Tuple
└──      return %3

```

**Comment by cscherrer** [comment]
Yeah, the reason I started down this path was that statically-typed things should have better performance than I was able to get out of `Recursive`. But maybe there's a way to fix that more directly, I'm not sure
